Core/include/Acts/Geometry/detail/NavigationStateUpdators.hpp

```diff
@@ -0,0 +1,173 @@
+
+#include <array>
+#include <memory>
+
+namespace Acts {
+namespace Experimental {
+namespace detail {
+
+/// @brief  This sets a single object, e.g. single surface or single volume
+/// @tparam object_type the type of the object to be filled
+/// @tparam filler_type is the helper to fill the object into nState
+template <typename object_type, typename filler_type>
+class SingleObjectImpl : public IManagedDelegateImpl {
+ public:
+  // The object to be filled in
+  const object_type* object = nullptr;
+
+  /// Convenience constructor
+  /// @param so the single object
+  SingleObjectImpl(const object_type* so) : object(so) {}
+
+  /// @brief updates the navigation state with a single object that is filled in
+  ///
+  /// @param gctx is the Geometry context of this call
+  /// @param nState the navigation state to which the surfaces are attached
+  ///
+  /// @note this is attaching objects without intersecting nor checking
+  void update([[maybe_unused]] const GeometryContext& gctx,
+              NavigationState& nState) const {
+    filler_type::fill(nState, object);
+  }
+};
+
+/// @brief This uses state less extractor and fillers to manipulate
+/// the navigation state
+///
+/// @tparam extractor_type the helper to extract the objects from
+/// @tparam filler_type is the helper to fill the object into nState
+template <typename extractor_type, typename filler_type>
+class StaticUpdatorImpl : public IManagedDelegateImpl {
+ public:
+  /// @brief updates the navigation state with a single object that is filled in
+  ///
+  /// @param gctx is the Geometry context of this call
+  /// @param nState the navigation state to which the surfaces are attached
+  ///
+  /// @note this is attaching objects without intersecting nor checking
+  void update([[maybe_unused]] const GeometryContext& gctx,
+              NavigationState& nState) const {
+    auto extracted = extractor_type::extract(gctx, nState);
+    filler_type::fill(nState, extracted);
+  }
+};
+
+/// @brief  This is an index grid based navigation state updator, it uses
+/// an extractor type and a filler type to handle the navigation state
+///
+/// @note a transform is applied `p3l = transform * p3` in order to allow
+/// shifted, transformed grids
+///
+/// It can be used for volumes, surfaces at convenience
+///
+/// @tparam grid_type is the type of the grid
+/// @tparam extractor_type is the helper to extract the object
+/// @tparam filler_type is the helper to fill the object into the nState
+template <typename grid_type, typename extractor_type, typename filler_type>
+class IndexedUpdatorImpl : public IManagedDelegateImpl {
+ public:
+  /// The grid where the indices are stored
+  grid_type grid;
+  /// An extractor helper to get the object(s) from the volume
+  extractor_type extractor;
+  /// These are the cast parameters
+  std::array<BinningValue, grid_type::DIM> casts;
+  /// A transform to be applied to the position
+  Transform3 transform = Transform3::Identity();
```

Comment:
Do these have to be public?

Reply:
Partly, I moved to private where possible.